### ISSUE TITLE
Update S3881: Correct documentation and add more unit tests

### ISSUE THREAD

**Issue by shadow-cs** [SonarSource/sonar-csharp#655]
### Description

SonarLint detects the following code as invalid `IDisposable` pattern implementation even though it is either generated by VisualStudio or copied directly from Sonar documentation.

SonarQube seems to not detect the issue only SonarLint does. 

May be related to #530.

### Repro steps

```c#
// Generated by VS
public class Foo1 : IDisposable
{

    #region IDisposable Support
    private bool disposedValue = false; // To detect redundant calls

    protected virtual void Dispose(bool disposing)
    {
        if (!disposedValue)
        {
            if (disposing)
            {
                // TODO: dispose managed state (managed objects).
            }

            disposedValue = true;
        }
    }

    // This code added to correctly implement the disposable pattern.
    public void Dispose()
    {
        // Do not change this code. Put cleanup code in Dispose(bool disposing) above.
        Dispose(true);
    }
    #endregion
}

// Taken from Sonar docs
public class Foo2 : IDisposable
{
    public void Dispose()
    {
        Dispose(true);
    }

    protected virtual void Dispose(bool disposing)
    {
        // Cleanup
    }
}
```
### Expected behavior

No issue detected.

### Actual behavior

Both `Foo1` and `Foo2` are detected as violation of proper `IDisposable` implementation.

`GC.SuppressFinalize(this);` seems to be required now. This may be a documentation issue as it seems the requirements for S3881 has changed recently.

### Known workarounds

Suppress the warning manually.

### Related information

* SonarC# 3.4.0.1732
* Visual Studio 17.2

**Comment by Evangelink** [comment]
Hi @shadow-cs, are you sure you are using SonarC# 3.4.0.1732? This is the version of the latest released SonarLint for Visual Studio. Are you using the standalone or the connected mode?

**Comment by michalb-sonar** [comment]
`GC.SuppressFinalize(this)` is necessary when the class has a finalizer. Here, the class does not have one, however it is unsealed, so it may be present in child classes.

Microsoft recommends to add SuppressFinalize in that case (despite of VS suggestion not following it).: https://docs.microsoft.com/en-us/dotnet/api/system.idisposable?view=netframework-4.7

Thanks for a good catch @shadow-cs! We'll update documentation. The rule will continue to enforce the presence of the call.
